cudamat/sparse.py

```diff
@@ -0,0 +1,101 @@
+
+import scipy.sparse as sp
+import cudamat
+import ctypes as ct
+
+import os
+import platform
+import warnings
+import sysconfig
+
+import ctypes as ct
+import numpy as np
+def load_library(basename):
+    if platform.system() == 'Windows':
+       ext = '.dll'
+    else:
+       ext = sysconfig.get_config_var('SO')
+    return ct.cdll.LoadLibrary(os.path.join(
+        os.path.dirname(__file__) or os.path.curdir,
+        basename + ext))
+
+
+_cudamat = load_library('libcudamat')
+
+_cudamat.sparse_dot.restype = ct.c_int
+
+class sparse_data(ct.Structure):
+    _fields_ = [('indptr', ct.POINTER(ct.c_int)),
+                ('indices', ct.POINTER(ct.c_int)),
+                ('data', ct.POINTER(ct.c_float))]
+
+class cudamat_sparse(ct.Structure):
+    _fields_ = [('data_host', sparse_data),
+                ('data_device', sparse_data),
+                ('on_device', ct.c_int),
+                ('on_host', ct.c_int),
+                ('size', ct.c_int * 2),
+                ('is_trans', ct.c_int),
+                ('owns_data', ct.c_int),
+                ('nnz', ct.c_int)]
+
+class SparseCUDAMatrix(object):
+    """ A SparseCUDAMatrix object represents a scipy.sparse.csr matrix of single
+    precision floats on a GPU.
+    """
+    def __init__(self, array, copy_to_device = True, copy_on_host=True):
+        """
+        Initializes a new matrix object in one of two ways. If array is a numpy
+        ndarray, memory for a matrix with the same dimensions is allocated on
+        the GPU. If the copy_to_device flag is set to True, the GPU matrix is
+        initialized with the given ndarray. If array is not an ndarray, it must
+        be a cudamat structure (typically the user will never use this way of
+        calling __init__).
```

Comment:
Needs updating. `array` is not a `ndarray` but a scipy CSR matrix, and `cuamat` structure is a `cudamat_sparse` structure. Furthermore, `copy_on_host` is not explained.